src/db_reader.rs

```diff
@@ -0,0 +1,816 @@
+            &path,
+            Arc::clone(&object_store),
+            Arc::clone(&options.clock),
+        ));
+
+        let table_store = Arc::new(TableStore::new(
+            Arc::clone(&object_store),
+            SsTableFormat::default(),
+            path.clone(),
+            options.block_cache.clone(),
+        ));
+
+        let mut manifest = StoredManifest::load(Arc::clone(&manifest_store)).await?;
+        if !manifest.db_state().initialized {
+            return Err(SlateDBError::InvalidDBState);
+        }
+
+        let checkpoint =
+            Self::get_or_create_checkpoint(&mut manifest, checkpoint_id, &options).await?;
+
+        let inner =
+            Arc::new(DbReaderInner::new(manifest_store, table_store, options, checkpoint).await?);
+
+        // If no checkpoint was provided, then we have established a new checkpoint
+        // from the latest state, and we need to refresh it according to the params
+        // of `DbReaderOptions`.
+        let manifest_poller = if checkpoint_id.is_none() {
+            Some(inner.spawn_manifest_poller()?)
+        } else {
+            None
+        };
+
+        Ok(Self {
+            inner,
+            manifest_poller,
+        })
+    }
+
+    async fn get_or_create_checkpoint(
+        manifest: &mut StoredManifest,
+        checkpoint_id: Option<Uuid>,
+        options: &DbReaderOptions,
+    ) -> Result<Checkpoint, SlateDBError> {
+        let checkpoint = if let Some(checkpoint_id) = checkpoint_id {
+            manifest
+                .db_state()
+                .find_checkpoint(&checkpoint_id)
+                .ok_or(SlateDBError::CheckpointMissing(checkpoint_id))?
+                .clone()
+        } else {
+            // Create a new checkpoint from the latest state
+            let options = CheckpointOptions {
+                lifetime: options.checkpoint_lifetime.clone(),
+                ..CheckpointOptions::default()
+            };
+            manifest.write_checkpoint(None, &options).await?
+        };
+        Ok(checkpoint)
+    }
+
+    /// Close the database reader.
+    ///
+    /// ## Returns
+    /// - `Result<(), SlateDBError>`: if there was an error closing the reader
+    ///
+    /// ## Examples
+    ///
+    /// ```
+    /// use slatedb::{DbReader, SlateDBError};
+    /// use slatedb::object_store::{ObjectStore, memory::InMemory};
+    /// use std::sync::Arc;
+    ///
+    /// #[tokio::main]
+    /// async fn main() -> Result<(), SlateDBError> {
+    ///     use slatedb::DbReaderOptions;let object_store: Arc<dyn ObjectStore> = Arc::new(InMemory::new());
+    ///     let options = DbReaderOptions::default();
+    ///     let reader = DbReader::open("test_db", object_store, None, options).await?;
+    ///     reader.close().await?;
+    ///     Ok(())
+    /// }
+    /// ```
+    ///
+    pub async fn close(&self) -> Result<(), SlateDBError> {
+        if let Some(poller) = &self.manifest_poller {
```

Comment:
Should any remaining checkpoints from this reader be removed if the reader was managing them (refreshing them)?

Reply:
The poller removes the checkpoint when it receives the `Shutdown`. I'll add an assertion in one of the test cases to verify that this works.